File: casatables/examples/write_ms.cpp
// Eqivalent C++ code for the write_ms.rs example
//
// e.g.
// ```bash
// cd casatables/examples
// make
// rm -rf /tmp/test.ms; strace -f -s 256 -k -o write_ms_cpp_create_init.strace ./write_ms -path /tmp/test.ms -write_mode create_only -initialize true
// rm -rf /tmp/test.ms; strace -f -s 256 -k -o write_ms_cpp_create_noinit.strace ./write_ms -path /tmp/test.ms -write_mode create_only -initialize false
// ```
// ```txt
//    5176 write_ms_cpp_create_init.strace
//    3848 write_ms_cpp_create_noinit.strace
// ```

#include <casacore/casa/Exceptions/Error.h>
#include <casacore/tables/Tables/Table.h>
#include <casacore/tables/Tables/TableDesc.h>
#include <casacore/tables/Tables/SetupNewTab.h>
#include <casacore/tables/Tables/ScalarColumn.h>
#include <casacore/tables/Tables/ArrayColumn.h>
#include <casacore/tables/Tables/ScaColDesc.h>
#include <casacore/tables/Tables/ArrColDesc.h>
#include <casacore/casa/Arrays/Vector.h>
#include <casacore/casa/Arrays/Matrix.h>
#include <casacore/casa/Arrays/IPosition.h>
#include <casacore/casa/BasicSL/Complex.h>
#include <casacore/tables/DataMan/TSMOption.h>
#include <vector>
#include <complex>
#include <string>
#include <map>

// Function to parse command-line arguments
std::map<std::string, std::string> parse_args(int argc, char* argv[]) {
    std::map<std::string, std::string> args;
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg[0] == '-') {
            std::string key = arg.substr(1);
            std::string value = "";
            if ((i + 1) < argc && argv[i + 1][0] != '-') {
                value = argv[++i];
            }
            args[key] = value;
        }
    }
    return args;
}

int main(int argc, char* argv[]) {
    try {
        // Parse command-line arguments
        auto args = parse_args(argc, argv);

        // Extract options with defaults
        std::string table_path = args.count("path") ? args["path"] : "/tmp/write_ms_cpp.ms";
        casacore::uInt n_rows = args.count("rows") ? std::stoi(args["rows"]) : 100;
        std::string tsm_option_str = args.count("tsm_option") ? args["tsm_option"] : "DEFAULT";
        bool initialize = args.count("initialize") ? (args["initialize"] == "true") : false;
        std::string write_mode = args.count("write_mode") ? args["write_mode"] : "create_only";
        std::string data_shape_str = args.count("data_shape") ? args["data_shape"] : "32,4";

        // Parse data shape
        std::vector<casacore::uInt> data_shape;
        std::stringstream ss(data_shape_str);
        std::string item;
        while (std::getline(ss, item, ',')) {
            data_shape.push_back(std::stoi(item));
        }
        casacore::IPosition data_shape_ipos(2, data_shape[0], data_shape[1]);

        // Map TSM option
        casacore::TSMOption::Option opt = casacore::TSMOption::Default;
        if (tsm_option_str == "MMAP") opt = casacore::TSMOption::MMap;
        else if (tsm_option_str == "BUFFER") opt = casacore::TSMOption::Buffer;
        else if (tsm_option_str == "CACHE") opt = casacore::TSMOption::Cache;
        else if (tsm_option_str == "AIPSRC") opt = casacore::TSMOption::Aipsrc;
        else if (tsm_option_str == "DEFAULT") opt = casacore::TSMOption::Default;
        else {
            std::cerr << "Unknown TSM option: " << tsm_option_str << std::endl;
            return 1;
        }

        // print all the args
        std::cout << "table_path: " << table_path << std::endl;
        std::cout << "n_rows: " << n_rows << std::endl;
        std::cout << "tsm_option_str: " << tsm_option_str << std::endl;
        std::cout << "initialize: " << initialize << std::endl;
        std::cout << "write_mode: " << write_mode << std::endl;
        std::cout << "data_shape_str: " << data_shape_str << std::endl;

        casacore::TSMOption tsmOpt(opt);

        // Create table description
        casacore::TableDesc td("test", "1", casacore::TableDesc::Scratch);
        td.addColumn(casacore::ScalarColumnDesc<casacore::Double>("TIME", "Observation time"));
        td.addColumn(casacore::ScalarColumnDesc<casacore::Int>("ANTENNA1", "First antenna"));
        td.addColumn(casacore::ScalarColumnDesc<casacore::Int>("ANTENNA2", "Second antenna"));
        td.addColumn(casacore::ScalarColumnDesc<casacore::Bool>("FLAG_ROW", "Row flag"));
        td.addColumn(casacore::ArrayColumnDesc<casacore::Complex>("DATA", "Visibility data", data_shape_ipos, casacore::ColumnDesc::FixedShape));
        td.addColumn(casacore::ArrayColumnDesc<casacore::Bool>("FLAG", "Data flags", data_shape_ipos, casacore::ColumnDesc::FixedShape));

        // Create the table
        std::cout << "Creating SetupNewTable..." << std::endl;
        casacore::SetupNewTable setup(table_path, td, casacore::Table::New);
        std::cout << "Creating Table..." << std::endl;
        casacore::Table table(setup, casacore::Table::Plain, n_rows, initialize, casacore::Table::LocalEndian, tsmOpt);

        // Write data based on mode
        if (write_mode == "create_only") {
            // Do nothing
        } else if (write_mode == "table_put_cell") {
            casacore::ScalarColumn<casacore::Double> time_col(table, "TIME");
            casacore::ScalarColumn<casacore::Int> ant1_col(table, "ANTENNA1");
            casacore::ScalarColumn<casacore::Int> ant2_col(table, "ANTENNA2");
            casacore::ScalarColumn<casacore::Bool> flag_row_col(table, "FLAG_ROW");
            casacore::ArrayColumn<casacore::Complex> data_col(table, "DATA");
            casacore::ArrayColumn<casacore::Bool> flag_col(table, "FLAG");

            casacore::Matrix<casacore::Complex> data_matrix(data_shape_ipos);
            casacore::Matrix<casacore::Bool> flag_matrix(data_shape_ipos, false);
            for (casacore::uInt i = 0; i < data_shape_ipos[0]; ++i) {
                for (casacore::uInt j = 0; j < data_shape_ipos[1]; ++j) {
                    casacore::uInt idx = i * data_shape_ipos[1] + j;
                    data_matrix(i, j) = casacore::Complex(static_cast<float>(idx), 0.0f);
                    flag_matrix(i, j) = (idx % 13 == 0);
                }
            }

            for (casacore::uInt row_idx = 0; row_idx < n_rows; ++row_idx) {
                time_col.put(row_idx, static_cast<casacore::Double>(row_idx));
                ant1_col.put(row_idx, static_cast<casacore::Int>(row_idx % 128));
                ant2_col.put(row_idx, static_cast<casacore::Int>((row_idx + 1) % 128));
                flag_row_col.put(row_idx, (row_idx % 2 == 0));

                data_col.put(row_idx, data_matrix);
                flag_col.put(row_idx, flag_matrix);
            }
        } else {
            std::cerr << "Unknown write mode: " << write_mode << std::endl;
            return 1;
        }

        // Clean up
        // table.markForDelete();
        // std::filesystem::remove_all(std::filesystem::path(table_path));

        std::cout << "C++ syscall tracer with casacore completed successfully." << std::endl;

    } catch (const casacore::AipsError& e) {
        std::cerr << "CasaCore error: " << e.getMesg() << std::endl;
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "Standard error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: casatables/examples/write_ms.rs
use anyhow::Error;
use clap::{Arg, Command};
use ndarray::Array2;
use rubbl_casatables::{
    Complex, GlueDataType, TSMOption, Table, TableCreateMode, TableDesc, TableDescCreateMode,
};
use rubbl_core::{ctry, notify::ClapNotificationArgsExt};
use std::{path::PathBuf, process, str::FromStr};

/// Perform basic casatables operations on a mock Measurement Set
///
/// Example usage with strace:
/// ```bash
/// cd casatables/examples
/// cargo build --example write_ms
/// rm -rf /tmp/test.ms; strace -f -s 256 -k -o write_ms_rust_create_init.strace ../../target/debug/examples/write_ms /tmp/test.ms -w create_only -i true
/// rm -rf /tmp/test.ms; strace -f -s 256 -k -o write_ms_rust_create_noinit.strace ../../target/debug/examples/write_ms /tmp/test.ms -w create_only -i false
/// wc -l *.strace
/// ```
/// ```txt
///    7949 write_ms_rust_create_init.strace
///    5871 write_ms_rust_create_noinit.strace
/// ```
///
fn main() {
    let matches = Command::new("benchmark")
        .rubbl_notify_args()
        .arg(
            Arg::new("TABLE-PATH")
                .value_parser(clap::value_parser!(PathBuf))
                .help("The path where the benchmark table will be created")
                .default_value("/tmp/write_ms_rust.ms")
                .index(1),
        )
        .arg(
            Arg::new("rows")
                .short('r')
                .long("rows")
                .value_parser(clap::value_parser!(usize))
                .help("Number of rows to create")
                .default_value("100"),
        )
        .arg(
            Arg::new("tsm_option")
                .short('t')
                .long("tsm_option")
                .value_parser(["CACHE", "BUFFER", "MMAP", "DEFAULT", "AIPSRC"])
                .help("TSM option to use")
                .default_value("DEFAULT"),
        )
        .arg(
            Arg::new("initialize")
                .short('i')
                .long("initialize")
                .help("Initialize the table")
                .value_parser(clap::value_parser!(bool))
                .default_value("false"),
        )
        .arg(
            Arg::new("write_mode")
                .short('w')
                .long("write_mode")
                .value_parser(["create_only", "table_put_cell"])
                .help("Write mode to use")
                .default_value("table_put_cell"),
        )
        .arg(
            Arg::new("data_shape")
                .short('d')
                .long("data_shape")
                .value_parser(|s: &str| {
                    s.split(',')
                        .map(|x| x.trim().parse::<u64>())
                        .collect::<Result<Vec<_>, _>>()
                })
                .help("Data shape to use (comma-separated, e.g. 32,4)")
                .default_value("32,4"),
        )
        .get_matches();

    process::exit(rubbl_core::notify::run_with_notifications(
        matches,
        |matches, _nbe| -> Result<i32, Error> {
            let table_path = matches.get_one::<PathBuf>("TABLE-PATH").unwrap();
            let n_rows = *matches.get_one::<usize>("rows").unwrap();
            let tsm_option =
                TSMOption::from_str(matches.get_one::<String>("tsm_option").unwrap()).unwrap();
            let initialize = matches.get_one::<bool>("initialize").unwrap();
            let write_mode = matches.get_one::<String>("write_mode").unwrap();
            let data_shape = matches.get_one::<Vec<u64>>("data_shape").unwrap();

            let mut table: Table = create_test_table(
                table_path,
                n_rows,
                data_shape.as_slice(),
                *initialize,
                Some(tsm_option),
            )?;
            let data_template = create_test_data(data_shape.as_slice());
            let flags_template = create_test_flags(data_shape.as_slice());

            match write_mode.as_str() {
                "create_only" => {
                    // Do nothing
                }
                "table_put_cell" => {
                    write_test_data_table_put_cell(&mut table, &data_template, &flags_template)?;
                }
                _ => {
                    eprintln!("Unknown write mode '{}'", write_mode);
                    std::process::exit(1);
                }
            }

            Ok(0)
        },
    ));
}

fn create_test_table(
    table_path: &PathBuf,
    n_rows: usize,
    data_shape: &[u64],
    initialize: bool,
    tsm_option: Option<TSMOption>,
) -> Result<Table, Error> {
    // Create a fresh table using the rubbl API that mirrors the C++ demo
    // Build the same schema as the C++ example using existing rubbl APIs
    let mut table_desc = ctry!(TableDesc::new("syscall_test", TableDescCreateMode::TDM_NEW);
        "failed to create test table description"
    );
    // Scalars - match C++ exactly
    table_desc.add_scalar_column(
        GlueDataType::TpDouble,
        "TIME",
        Some("Observation time"),
        false,
        false,
    )?;
    table_desc.add_scalar_column(
        GlueDataType::TpInt,
        "ANTENNA1",
        Some("First antenna"),
        false,
        false,
    )?;
    table_desc.add_scalar_column(
        GlueDataType::TpInt,
        "ANTENNA2",
        Some("Second antenna"),
        false,
        false,
    )?;
    table_desc.add_scalar_column(
        GlueDataType::TpBool,
        "FLAG_ROW",
        Some("Row flag"),
        false,
        false,
    )?;
    // Fixed-shape arrays - match C++ exactly (no comments for arrays)
    // For now we always add arrays the same way; backend SM selection is handled by TSMOption
    table_desc.add_array_column(
        GlueDataType::TpComplex,
        "DATA",
        None,
        Some(data_shape),
        false,
        false,
    )?;
    table_desc.add_array_column(
        GlueDataType::TpBool,
        "FLAG",
        None,
        Some(data_shape),
        false,
        false,
    )?;

    let table = Table::new(
        table_path,
        table_desc,
        n_rows,
        TableCreateMode::New,
        initialize,
        tsm_option,
    )?;
    Ok(table)
}

fn create_test_data(data_shape: &[u64]) -> Array2<Complex<f32>> {
    let shape = (data_shape[0] as usize, data_shape[1] as usize);
    let mut data = Array2::<Complex<f32>>::zeros(shape);

    // Fill with some pattern to make it interesting for analysis
    for ((i, j), elem) in data.indexed_iter_mut() {
        *elem = Complex::new(
            (i as f32 * 0.1).sin() * (j as f32 * 0.2).cos(),
            (i as f32 * 0.15).cos() * (j as f32 * 0.25).sin(),
        );
    }

    data
}

fn create_test_flags(data_shape: &[u64]) -> ndarray::Array2<bool> {
    let shape = (data_shape[0] as usize, data_shape[1] as usize);
    let mut flags = ndarray::Array2::<bool>::from_elem(shape, false);

    // Set some flags to create realistic patterns
    for ((i, j), elem) in flags.indexed_iter_mut() {
        *elem = (i + j) % 17 == 0; // Arbitrary pattern
    }

    flags
}

fn write_test_data_table_put_cell(
    table: &mut Table,
    data_template: &Array2<Complex<f32>>,
    flags_template: &ndarray::Array2<bool>,
) -> Result<(), Error> {
    let rows_to_write = table.n_rows();
    for row_u64 in 0..rows_to_write {
        // Scalars
        table.put_cell("TIME", row_u64, &(row_u64 as f64))?;
        table.put_cell("ANTENNA1", row_u64, &((row_u64 as i32) % 128))?;
        table.put_cell("ANTENNA2", row_u64, &((row_u64 as i32 + 1) % 128))?;
        table.put_cell("FLAG_ROW", row_u64, &((row_u64 % 2) == 0))?;

        // Arrays
        table.put_cell("DATA", row_u64, data_template)?;
        table.put_cell("FLAG", row_u64, flags_template)?;
    }
    Ok(())
}

// No readback: benchmark only creates and writes the table
